packages/agreement-process/src/services/agreementService.ts

```diff
@@ -277,3 +347,163 @@ export async function updateAgreementLogic({
   );
 }
+
+// eslint-disable-next-line sonarjs/cognitive-complexity
+export async function upgradeAgreementLogic({
+  agreementId,
+  authData,
+  agreementQuery,
+  eserviceQuery,
+  tenantQuery,
+  fileCopy,
+}: {
+  agreementId: string;
+  authData: AuthData;
+  agreementQuery: AgreementQuery;
+  eserviceQuery: EserviceQuery;
+  tenantQuery: TenantQuery;
+  fileCopy: (
+    container: string,
+    sourcePath: string,
+    destinationPath: string,
+    destinationFileName: string,
+    docName: string
+  ) => Promise<string>;
+}): Promise<{ streamId: string; events: Array<CreateEvent<AgreementEvent>> }> {
+  const agreementToBeUpgraded = await agreementQuery.getAgreementById(
+    agreementId
+  );
+  const tenant = await tenantQuery.getTenantById(authData.organizationId);
+  assertTenantExist(authData.organizationId, tenant);
+  assertAgreementExist(agreementId, agreementToBeUpgraded);
+  assertRequesterIsConsumer(
+    agreementToBeUpgraded.data.consumerId,
+    authData.organizationId
+  );
+
+  assertExpectedState(
+    agreementId,
+    agreementToBeUpgraded.data.state,
+    agreementUpgradableStates
+  );
+
+  const eservice = await eserviceQuery.getEServiceById(
+    agreementToBeUpgraded.data.eserviceId
+  );
+  assertEServiceExist(agreementToBeUpgraded.data.eserviceId, eservice);
+
+  const newDescriptor = eservice.data.descriptors.find(
+    (d) => d.state === descriptorState.published
+  );
+  if (newDescriptor === undefined) {
+    throw publishedDescriptorNotFound(agreementToBeUpgraded.data.eserviceId);
+  }
+  const latestDescriptorVersion = z
+    .preprocess((x) => Number(x), z.number())
+    .safeParse(newDescriptor.version);
+  if (!latestDescriptorVersion.success) {
+    throw unexpectedVersionFormat(eservice.data.id, newDescriptor.id);
+  }
+
+  const currentDescriptor = eservice.data.descriptors.find(
+    (d) => d.id === agreementToBeUpgraded.data.descriptorId
+  );
+  if (currentDescriptor === undefined) {
+    throw descriptorNotFound(
+      eservice.data.id,
+      agreementToBeUpgraded.data.descriptorId
+    );
+  }
+
+  const currentVersion = z
+    .preprocess((x) => Number(x), z.number())
+    .safeParse(currentDescriptor.version);
+  if (!currentVersion.success) {
+    throw unexpectedVersionFormat(eservice.data.id, currentDescriptor.id);
+  }
+
+  if (latestDescriptorVersion.data <= currentVersion.data) {
+    throw noNewerDescriptor(eservice.data.id, currentDescriptor.id);
+  }
+
+  if (eservice.data.producerId !== authData.organizationId) {
+    validateCertifiedAttributes(newDescriptor, tenant.data);
+  }
+
+  const verifiedValid = verifiedAttributesSatisfied(
+    agreementToBeUpgraded.data.producerId,
+    newDescriptor,
+    tenant.data
+  );
+
+  const declaredValid = declaredAttributesSatisfied(newDescriptor, tenant.data);
+
+  if (verifiedValid && declaredValid) {
+    // upgradeAgreement
+    const stamp: AgreementStamp = {
+      who: authData.organizationId,
+      when: new Date(),
```

Comment:
@Viktor-K here I think we have the same issues discussed in: https://github.com/pagopa/interop-be-monorepo/pull/136#discussion_r1439500454

Should probably be fixed before merging?

Reply:
I suggest to leave this as is, after this PR and others related is closed we will adopt [timestampUtils](https://github.com/pagopa/interop-be-monorepo/pull/140/files#diff-588f8f0554d02253ed100e28acf0452df95a6d23bce1133fec6e2393d551535e) in every AgreemetStamp creation.